### PR TITLE
[JENKINS-35234] Support the new workflow/pipeline model

### PR DESCRIPTION
**Pull request by piomis**
pom.xml
Added dependencies to workflow modules
Sloccount plugin now requires Jenkins 2.63

Refactored for new workflow (AbstractBuild and AbstractProject replaced
with Run<> and Job<>)
SloccountAreaRenderer.java
SloccountBuildAction.java
SloccountProjectAction.java
SloccountPublisher.java
SloccountResult.java

Solved NPE:
SloccountResult.java

Ignored FindBug errors:
SloccountAreaRenderer.java

Solved javadoc complains:
SloccountTablePortlet.java
SloccountLanguageStatistics.java
SloccountParser.java
ClocFile.java
ClocFiles.java
ClocHeader.java
ClocReport.java
ClocTotal.java

Updated API for new Jenkins:
FileFinder.java

Added escape-by-default in all jelly files:
index.jelly
SloccountTablePortlet/portlet.jelly
SloccountTablePortlet/table.jelly
SloccountBuildAction/statistics.jelly
SloccountBuildAction/summary.jelly
SloccountProjectAction/floatingBox.jelly
SloccountProjectAction/jobMain.jelly
SloccountPublisher/config.jelly
SloccountResult/files.jelly
SloccountResult/folders.jelly
SloccountResult/index.jelly
SloccountResult/languages.jelly
SloccountResult/modules.jelly
tabview/css.jelly
tabview/distribution-graph.jelly
tabview/main.jelly

Added SloccountPublisher step supporting workflow (with jelly file and
tests)
steps/SloccountPublisherStep.java
SloccountPublisherStep/config.jelly
test/SloccountPublisherStepTest.java

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers some reviews would be useful here

File: pom.xml [hunk 1]
```diff
@@ -55,11 +55,49 @@
 
   <dependencies>
     <dependency>
+     <groupId>org.jenkins-ci.main</groupId>
```
Comment (oleg-nenashev) [inline pom.xml]: It should not be done in such way. Effectively you bundle Jenkins core and its deps into a plugin, which may cause some inexpected results when a user installs the plugin.

If you need a specific core version, use

```
<properties>
  <jenkins.version>1.642.3</jenkins.version>
</properties>
```

File: pom.xml [hunk 2]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.447</version>
+    <version>2.7</version>
```
Comment (oleg-nenashev) [inline pom.xml]: Better to use the latest version

File: pom.xml [hunk 3]
```diff
@@ -55,11 +55,49 @@
...
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>dashboard-view</artifactId>
       <version>2.0</version>
       <optional>true</optional>
     </dependency>
+    <!-- dependencies on Jenkins Pipeline plugins -->  
+    <dependency>  
+      <groupId>org.jenkins-ci.plugins.workflow</groupId>  
+      <artifactId>workflow-job</artifactId>  
+      <version>1.14</version>
```
Comment (oleg-nenashev) [inline pom.xml]: It effectively makes the dependency on Pipeline mandatory. Which is not what particular Jenkins admins want.
If you need it for testing only, use `<scope>test</scope>`.

Same applies to the dependencies below

File: src/main/java/hudson/plugins/sloccount/SloccountAreaRenderer.java [hunk 4]
```diff
@@ -10,6 +11,7 @@
  * links. This renderer does not render tooltips, these need to be defined in
  * sub-classes.
  */
+@SuppressFBWarnings(value="EQ_DOESNT_OVERRIDE_EQUALS", justification="Equals method is not needed.")
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountAreaRenderer.java]: Not sure about it, seems to be a valid bug

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java [hunk 5]
```diff
@@ -101,11 +115,33 @@ SloccountBuildAction getPreviousAction(){
...
         return this.result;
     }
+    
+    @Override
+    public void onLoad(Run<?,?> r)
+    {
+        this.build = r;
+        if( this.projectActions == null )
+        {
+            this.projectActions = new ArrayList<>();
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: This collection type is not thread-safe. Without synchronization such logic will likely fail at some point

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java [hunk 6]
```diff
@@ -101,11 +115,33 @@ SloccountBuildAction getPreviousAction(){
         return null;
     }
 
-    public AbstractBuild<?,?> getBuild(){
+    public Run<?,?> getBuild(){
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: It breaks the binary compatibility. But it's up to the plugin maintainer

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java [hunk 7]
```diff
@@ -37,6 +45,12 @@ public String getDisplayName() {
     public String getUrlName() {
         return URL_NAME;
     }
+    
+    @Override  
+    public Collection<? extends Action> getProjectActions() {  
+        return this.projectActions;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: From what I see you could just generate this list on the flight. No need to cache it

File: src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java [hunk 8]
```diff
@@ -1,36 +1,33 @@
...
 import org.kohsuke.stapler.StaplerResponse;
 
 /**
  *
  * @author lordofthepigs
  */
-public class SloccountProjectAction implements Action, Serializable {
-    /** Serial version UID. */
-    private static final long serialVersionUID = 0L;
+public class SloccountProjectAction implements Action {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java]: This action has never been serializable, so I'd guess it's fine to remove it

File: src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java [hunk 9]
```diff
@@ -75,8 +72,8 @@ public void doIndex(final StaplerRequest request, final StaplerResponse response
      * @return the last finished build or <code>null</code> if there is no
      *         such build
      */
-    public AbstractBuild<?, ?> getLastFinishedBuild() {
-        AbstractBuild<?, ?> lastBuild = project.getLastBuild();
+    public Run<?, ?> getLastFinishedBuild() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java]: incompatible change

File: src/main/java/hudson/plugins/sloccount/SloccountResult.java [hunk 10]
```diff
@@ -25,7 +25,7 @@
 
     private transient SloccountReport report;
 
-    private final AbstractBuild<?,?> owner;
+    private transient final Run<?,?> owner;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountResult.java]: I am not sure this change is enough.
The `owner` field won't be persisted to the disk anymore, and I do not see logic where it gets restored after loading from the disk. Maybe it's not required

File: src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java [hunk 11]
```diff
@@ -23,6 +23,9 @@
 
     /**
      * Constructor.
+     * @param blank blank
+     * @param comment comment
+     * @param code code
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java]: Comments just to have comments. Not really required

File: src/main/java/hudson/plugins/sloccount/util/FileFinder.java [hunk 12]
```diff
@@ -9,6 +9,7 @@
 import org.apache.tools.ant.BuildException;
 import org.apache.tools.ant.Project;
 import org.apache.tools.ant.types.FileSet;
+import org.jenkinsci.remoting.RoleChecker;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/util/FileFinder.java]: Should be MasterToSlaveFileCallable instead. Or something like that

File: src/main/java/jenkins/plugins/sloccount/steps/SloccountPublisherStep.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution;
+
+/**
+ *
+ * @author piomis
+ */
+public class SloccountPublisherStep extends Step implements Serializable {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/sloccount/steps/SloccountPublisherStep.java]: You have already converted `SloccountPublisher` to `SimpleBuildStep`. It should be enough, no?

File: src/main/java/hudson/plugins/sloccount/SloccountAreaRenderer.java (hunk 4 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountAreaRenderer.java]: Honestly - it looks that new "maven" includes FB checks. I did not find any way to fix it (lots of other I fixed).
In old environment there was no complain for this piece of code.

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java (hunk 6 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: I followed one of blog post (I think it was for Gatling plugin) - there was info that AbstractBuild and AbstractProject are converted to Run and Job.

File: src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java (hunk 8 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java]: This was exactly what I meant. It looks that ProjectActions are not saved.
Only BuildActions are.

File: src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java (hunk 9 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountProjectAction.java]: I followed one of blog post (I think it was for Gatling plugin) - there was info that AbstractBuild and AbstractProject are converted to Run and Job.

File: src/main/java/hudson/plugins/sloccount/SloccountResult.java (hunk 10 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountResult.java]: As far as I recall I added transient due to FB.

File: src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java (hunk 11 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java]: Honestly there was no comment before. But older environment for build did not check JavaDoc results. New one complains for it - I did not find out all meaning for all parameters.

File: src/main/java/jenkins/plugins/sloccount/steps/SloccountPublisherStep.java (hunk 13 above)
Comment (piomis) [inline src/main/java/jenkins/plugins/sloccount/steps/SloccountPublisherStep.java]: I followed blog post (for Gantling plugin) which described how to add Pipeline support in plugin. There was also such class added - as far as I recall it was mainly to add "special function" in Groovy.
In this case "publishSloccount()"

File: pom.xml (hunk 2 above)
Comment (piomis) [inline pom.xml]: Which is?

File: src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java (hunk 11 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java]: > But older environment for build did not check JavaDoc results. New one complains for it - I did not find out all meaning for all parameters.

It prints warnings, but it does not block the build/release. So it is probably fine to skip them. It is also fine to leave the current change IMHO.

File: pom.xml (hunk 2 above)
Comment (oleg-nenashev) [inline pom.xml]: 2.31: https://github.com/jenkinsci/plugin-pom/releases

File: pom.xml (hunk 3 above)
Comment (piomis) [inline pom.xml]: I will check.
In fact most of dependecies was a result of experiments.
I will try to play more with dependecies.
However if plugin adds functionality to pipeline type of job shouldn't it have dependency to pipeline?

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java (hunk 5 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: This contruction I add after lots of fights with NPE after Jenkins restart.
Normally constructor is called and list is created. Later on onAttached is called and ProjectAction is added.
However after Jenkins shutdown and restart build is retrieved from disk.
In that case constructor is not called and onLoad raised NPE when I tried to add action to it.
As a result Action is not displayed at main Job page.

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java (hunk 7 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: Honestly I do not know how many time and whet getProjectActions can be called.
So to not create list and ProjectActions to many time I decided to create and add it ones on onLoad/onAttached and just return list on getProjectActions.

File: src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java (hunk 11 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java]: I will double check - but I am nearly 100% sure that build was stopped.

File: src/main/java/hudson/plugins/sloccount/SloccountResult.java (hunk 10 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/SloccountResult.java]: I have checked - it must be transient as Run<> is not serializable.
However field is needed due as it is referenced from jelly file (side panel).
I will add setter and properly initialize it from onLoad in BuildAction

File: pom.xml (hunk 2 above)
Comment (piomis) [inline pom.xml]: I changed to 2.30 - with 2.31 there was so complain during build

File: pom.xml (hunk 3 above)
Comment (piomis) [inline pom.xml]: I have developed new step (publishSloccount) thus dependency to workflow-step-api, workflow-job and workflow-cps is needed.
workflow-support and workflow-basic-steps are indeed for test purpose only.

File: src/main/java/hudson/plugins/sloccount/util/FileFinder.java (hunk 12 above)
Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/util/FileFinder.java]: I think that RoleChecked should stay (it is type of checkRoles argument).
However it looks that FilePath.FileCallable (imported few lines above) shall be replaced by  MasterToSlaveFileCallable , shan't it?

Comment (piomis) [inline src/main/java/hudson/plugins/sloccount/util/FileFinder.java]: After deeper investigation I believe that current implementation is fine.
FileFinder is used only by SloccountParser which already implements MasterToSlaveFileCallable.
I think that this should be enough (but I may be wrong).

**Comment by piomis** [comment]
@oleg-nenashev please re-review changes and my comments (and potentially unblock PR)

@jenkinsci/code-reviewers any other remarks or opinions would be very useful

File: src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java (hunk 11 above)
Comment (MarkEWaite) [reply src/main/java/hudson/plugins/sloccount/model/cloc/ClocTotal.java]: The JDK 8 javadoc warnings are not worth fighting.  It is much easier to insert the comments than force yourself to remember on each build that you can ignore that warning.

**Comment by piomis** [comment]
@oleg-nenashev @jenkinsci/code-reviewers just a ping :)

**Comment by piomis** [comment]
@oleg-nenashev @jenkinsci/code-reviewers just a ping :)

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java (hunk 6 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: It's still a breaking change. So I would advice a major version update in this case

File: pom.xml [hunk 14]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.447</version>
+    <version>2.30</version>
```
Comment (oleg-nenashev) [inline pom.xml]: Would be better to pick 2.35 now. There were some issues with Restricted checks in 2.30

File: src/main/java/hudson/plugins/sloccount/SloccountAreaRenderer.java (hunk 4 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/plugins/sloccount/SloccountAreaRenderer.java]: Then add a TODO comment which says "should be fixed". And probably link comment to a Jenkins ticket

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java [hunk 15]
```diff
@@ -1,29 +1,37 @@
...
 import org.kohsuke.stapler.StaplerProxy;
 
 /**
  *
  * @author lordofthepigs
  */
-public class SloccountBuildAction implements Action, Serializable, StaplerProxy {
+public class SloccountBuildAction implements RunAction2, StaplerProxy, SimpleBuildStep.LastBuildAction  {
     /** Serial version UID. */
-    private static final long serialVersionUID = 0L;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: the comment above should be removed as well then. I would recommend double checking that serialization is not required anywhere by this plugin. But I would not expect Actions to be really serializable

File: src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java (hunk 5 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java]: Still does not clarify the thread safety concern. What does block you from making methods more reliable if you really want to use internal caching?

**Comment by oleg-nenashev** [comment]
Sorry for the response delays. I am drowning in action items, and I cannot commit to respond quickly (or to at least read every notification in my inbox).

**Comment by hummerstudio** [comment]
@oleg-nenashev hello, when will we release the new version with this PR code? I am working on moving projects to Pipeline and I need this feature.

**Comment by oleg-nenashev** [comment]
@hummerstudio I am not even a maintainer of this plugin :( I am just interested in it and hence I am trying to contribute by reviews. But generally this plugin needs somebody to step forward and to take ownership of the plugin.

**Comment by ivaivalous** [comment]
I've created a new pull request to address the requested changes in this one, as it seems it has stalled: https://github.com/jenkinsci/sloccount-plugin/pull/49

@jenkinsci/code-reviewers, @oleg-nenashev, @piomis I would be happy if I can have your feedback.

**Pull request merged**